### PR TITLE
HEEDLS-476 Login Claims bug fix

### PR DESCRIPTION
**Pull request by ibrahimmunir14**
Refactor GetClaimsForSignIn into helper class; fix bug causing error when user has email=NULL; update which claims are set depending on user type; add some unit tests.

### CODE REVIEW

File: DigitalLearningSolutions.Web.Tests/Helpers/LoginClaimsHelperTests.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+{
+    using System.Security.Claims;
+    using DigitalLearningSolutions.Data.Tests.TestHelpers;
+    using DigitalLearningSolutions.Web.Controllers;
+    using DigitalLearningSolutions.Web.Helpers;
+    using DigitalLearningSolutions.Web.Models;
+    using FluentAssertions;
+    using NUnit.Framework;
+
+    public class LoginHelperTests
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Web.Tests/Helpers/LoginClaimsHelperTests.cs:11]: Nice tests, though could you just format them with Given-When-Then for ease of reading?

File: DigitalLearningSolutions.Web.Tests/Helpers/LoginClaimsHelperTests.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        [Test]
+        public void Delegate_user_forename_and_surname_set_correctly()
+        {
+            var delegateUser = UserTestHelper.GetDefaultDelegateUser(firstName: "fname", lastName: "lname");
+            var delegateLoginDetails = new DelegateLoginDetails(delegateUser);
+            var claims = LoginClaimsHelper.GetClaimsForSignIn(null, delegateLoginDetails);
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Web.Tests/Helpers/LoginClaimsHelperTests.cs:18]: When

File: DigitalLearningSolutions.Web.Tests/Helpers/LoginClaimsHelperTests.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            claims.Should().Contain((claim) => claim.Type == CustomClaimTypes.UserForename);
+            claims.Should().Contain((claim) => claim.Type == CustomClaimTypes.UserSurname);
+            var forenameClaim = claims.Find((claim) => claim.Type == CustomClaimTypes.UserForename);
+            var surnameClaim = claims.Find((claim) => claim.Type == CustomClaimTypes.UserSurname);
+            forenameClaim.Value.Should().Be("fname");
+            surnameClaim.Value.Should().Be("lname");
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Web.Tests/Helpers/LoginClaimsHelperTests.cs:24]: Then

File: DigitalLearningSolutions.Web.Tests/Helpers/LoginClaimsHelperTests.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    using FluentAssertions;
+    using NUnit.Framework;
+
+    public class LoginHelperTests
+    {
+        [Test]
+        public void Delegate_user_forename_and_surname_set_correctly()
+        {
+            var delegateUser = UserTestHelper.GetDefaultDelegateUser(firstName: "fname", lastName: "lname");
+            var delegateLoginDetails = new DelegateLoginDetails(delegateUser);
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Web.Tests/Helpers/LoginClaimsHelperTests.cs:17]: Given

File: DigitalLearningSolutions.Web.Tests/Helpers/LoginClaimsHelperTests.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            claims.Should().Contain((claim) => claim.Type == CustomClaimTypes.UserForename);
+            claims.Should().Contain((claim) => claim.Type == CustomClaimTypes.UserSurname);
+            var forenameClaim = claims.Find((claim) => claim.Type == CustomClaimTypes.UserForename);
+            var surnameClaim = claims.Find((claim) => claim.Type == CustomClaimTypes.UserSurname);
+            forenameClaim.Value.Should().Be("fname");
+            surnameClaim.Value.Should().Be("lname");
+        }
+        
+        [Test]
+        public void User_without_email_should_have_empty_string_email_claim()
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Web.Tests/Helpers/LoginClaimsHelperTests.cs:28]: You can shorten test names by removing `should` e.g. User_without_email_has_empty_string_email_claim.
Test names are often long, so shortening them is always nice, but it's not that big a deal 🤷

**Review by SteveJacksonSoft — CHANGES_REQUESTED** [review]
Looks nice, just a couple of small things worth tweaking.

File: DigitalLearningSolutions.Web/Helpers/LoginClaimsHelper.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            {
+                claims.Add(new Claim(CustomClaimTypes.UserAdminId, adminLoginDetails.Id.ToString()));
+            }
+
+            if (delegateLoginDetails?.Id != null)
+            {
+                claims.Add(new Claim(CustomClaimTypes.LearnCandidateId, delegateLoginDetails.Id.ToString()));
+            }
+
+            if (adminLoginDetails?.CategoryId != null)
```
[the remark this answers is not in the document]
Comment (ibrahimmunir14) [reply DigitalLearningSolutions.Web/Helpers/LoginClaimsHelper.cs:68]: Yeah that was my first instinct but didn't know if it would break something.

**Review by stellake — APPROVED** [review]
Looks good! 👍

**Pull request merged**